### DISCUSSION TITLE
NeLua Needs to be General and Independent

### DISCUSSION THREAD

**Discussion by ghost** [edubart/nelua-lang#254] [Ideas]
Lua is only known for being used as a scripting language in games like Roblox, Warcraft and Garry's Mod. But forget about Torch...

## NeLua needs to be the protagonist

In all the languages ​​I've seen and used, Lua has the most pleasant, minimalist and fun syntax to code. If Lua were not focused on being a language to extend other languages ​​(being used as a "second language"), and sought to be a general-purpose language like Python and Javascript, it would probably be one of the most used languages ​​in the programming world. And NeLua can do this, keeping the current features and minimalism of Lua, and adding necessary elements, like a regex library...

## NeLua should have a bigger community. Pancake package manager

Python has a giant package manager (PyPi) with incredible modules, such as for editing videos (moviepy), creating graphics (SciPy, Matplotlib), GUIs (Tk, Kivy), machine learning; And an impressive standard library. However, the Lua community does not have this wide range of tools in LuaRocks, for example, which attract beginner and enthusiast programmers. NeLua uses pancake and I'm rooting for it to be Lua's new "PyPi".

## NeLua needs an installer and an executable, like Python/Ruby/Lua Binaries

To make it easier to obtain NeLua, it would be interesting to make the compiled NeLua available on the Download page.

Here is an open discussion to talk about NeLua's protagonism. Thanks!

**Comment by asumbek** [comment]
1. Nelua is still unstable at the moment, and the development will probably take years until the release of 1.0. If Nelua is already used for production, I doubt you'd have that fun DX with Nelua and I cannot assure the software that's written in Nelua is really stable.
2. You don't need a package manager for that, `.neluacfg.lua` is simply the build system for Nelua. You can combine that with vcpkg or other C package managers. If that's not enough for you, create a dead-simple dependency manager, that hooks into `.neluacfg.lua`, that'll fetch the dependencies and make it available in the Nelua code upon building.
3. Did you mean prebuilt binaries? That's possible, we'd just wait Edubart for him to add Github workflows that automatically build Nelua and provide artifacts of the result build.

**Reply by Andre-LA** [reply]
About 3, that's the case already

https://github.com/edubart/nelua-lang/releases/tag/latest

**Reply by asumbek** [reply]
Nice to see that. But it would be better if it's packaged into a package manager repository. I think Nelua is already packaged for Arch Linux via AUR and Termux. Only most Linux distros' package repo, Windows' Scoop repo, and Homebrew's repo that doesn't contain Nelua yet.

**Reply by ghost** [reply]
About 2 and 3. It would be interesting if it were more accessible to beginner programmers, and easier to use.

Instead of downloading a C compiler, downloading vcpkg, compiling it to C, writing it in Lua and then running it, it would be much better to just run it directly. And install a package via a simple command (like Python's pip).
That's why Python, Go, Ruby and others are so widely used - the ease of getting it + simple code. If NeLua introduces the same facilities as Python (download, write, run) many programmers, both beginners and experts, may be interested in the language.
In short, to program in Lua, knowledge of C should not be necessary, except, of course, when you want to call C and Lua.

I make a lot of comparisons with Python due to its ease of handling, but in terms of code syntax and concepts, Lua is superior, so I want to bring the ease of running/installing Python to Lua's syntax.

**Reply by asumbek** [reply]
> Instead of downloading a C compiler

That's means we're forced to use LLVM, which is a pretty big dependency for a simple Lua-flavored C-compiled language. That would just be the same case as Rust or Zig, where they heavily depends on LLVM. LLVM itself is already slow for compiling a single simple code. That's why Nelua instead choose to transpile to C and let C compilers do the job for compiling instead, since it's much easier than having to rely on LLVM to do that.

> it would be much better to just run it directly

As if Nelua is an interpreted language. If you want an interpreted Nelua, I would recommend for you to use [Teal](https://github.com/teal-language/tl) instead.

> That's why Python, Go, Ruby and others are so widely used - the ease of getting it + simple code. If NeLua introduces the same facilities as Python (download, write, run) many programmers, both beginners and experts, may be interested in the language.

I have told you, Nelua is still unstable, and still young to actually have fights with other popular languages. And about the "download, write, run", we're actually trying to port Nelua into many package repositories. Nelua is already packaged in AUR, Termux. The rest of it are Linux distros such as Ubuntu, Windows via Chocolatey, Scoop or winget, and macOS via Homebrew. If it's already port to all of that, just simply use the package manager to install Nelua into the machine.
